wgpu-hal/src/vulkan/device.rs

```diff
@@ -531,8 +517,8 @@ impl
 }
 
 struct CompiledStage {
-    create_info: vk::PipelineShaderStageCreateInfo,
-    _entry_point: CString,
+    create_info: vk::PipelineShaderStageCreateInfo<'static>,
+    _entry_point: Pin<CString>,
```

Comment:
Why is why `Unpin` has never made sense to me, since `(C)String` contain a pointer internally. If I can perform destructive actions on the contents of `(C)String` (e.g. `mem::replace()`, or appending to the string causing a realloc), how can I guarantee the memory contents remain pinned?

Reply:
I looked at the docs and there is no way to mutate `CString`'s content. You can only do `mem::replace()` on the `CString` itself so the only way the pointer returned by `.as_ptr()` becomes invalid is if you drop the `CString`.

I find rust's safety docs to be top-notch and the docs of `.as_ptr()` indicate that the only invariant you have to care about is keeping the `CString` alive.

Reply:
Actually, never mind. In that example the `Pin` doesn't own the `String` but a borrow of the `String`, which is different. I think my understanding and interpretation of `Pin` still holds then. After all, `String` is safe to move around, the thing it points to internally is not.